File: mobile/native/arucon-widget/ios/AruconWidgetBridge/AruconWidgetBridgeModule.swift
import CoreFoundation
import ExpoModulesCore
import Foundation
import WidgetKit

private let appGroup = "group.com.arucon.dev.widget"
private let snapshotKey = "arucon.widget.snapshot.v1"
private let maximumSafeInteger: Int64 = 9_007_199_254_740_991
private let projectionKeys: Set<String> = [
  "petId", "stateRevision", "updatedAtMs", "formId", "personalityProfileId", "displayState"
]
private let displayStates: Set<String> = ["awake", "sleeping", "hibernating", "needs_care"]

private func nonemptyString(_ value: Any?) -> String? {
  guard let string = value as? String, !string.isEmpty else { return nil }
  return string
}

private func safeInteger(_ value: Any?) -> Bool {
  guard let number = value as? NSNumber,
        CFGetTypeID(number) != CFBooleanGetTypeID() else {
    return false
  }
  let integerEncodings: Set<String> = ["c", "s", "i", "l", "q", "C", "S", "I", "L", "Q"]
  guard integerEncodings.contains(String(cString: number.objCType)) else { return false }
  return number.int64Value >= 0 && number.int64Value <= maximumSafeInteger
}

private func validSnapshot(_ data: Data) -> Bool {
  guard let object = try? JSONSerialization.jsonObject(with: data),
        let dictionary = object as? [String: Any],
        Set(dictionary.keys) == projectionKeys,
        nonemptyString(dictionary["petId"]) != nil,
        nonemptyString(dictionary["formId"]) != nil,
        nonemptyString(dictionary["personalityProfileId"]) != nil,
        let displayState = nonemptyString(dictionary["displayState"]),
        displayStates.contains(displayState),
        safeInteger(dictionary["stateRevision"]),
        safeInteger(dictionary["updatedAtMs"]) else {
    return false
  }
  return true
}

public class AruconWidgetBridgeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AruconWidgetBridge")

    AsyncFunction("replaceSnapshotJsonAsync") { (snapshotJson: String) in
      guard let data = snapshotJson.data(using: .utf8), validSnapshot(data) else {
        throw Exception(name: "InvalidWidgetSnapshot", description: "Widget snapshot violates the six-field contract")
      }
      guard let defaults = UserDefaults(suiteName: appGroup) else {
        throw Exception(name: "WidgetContainerUnavailable", description: "Widget shared container is unavailable")
      }
      defaults.set(data, forKey: snapshotKey)
    }

    AsyncFunction("readSnapshotJsonAsync") { () -> String? in
      guard let defaults = UserDefaults(suiteName: appGroup),
            let data = defaults.data(forKey: snapshotKey),
            validSnapshot(data) else {
        return nil
      }
      return String(data: data, encoding: .utf8)
    }

    AsyncFunction("requestTimelineReloadAsync") { () -> String in
      WidgetCenter.shared.reloadTimelines(ofKind: "AruconWidget")
      return "requested"
    }
  }
}

File: mobile/native/arucon-widget/android/src/main/java/com/arucon/widgetbridge/AruconWidgetBridgeModule.kt
package com.arucon.widgetbridge

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject

private const val SNAPSHOT_PREFERENCES = "arucon.widget.snapshot.v1"
private const val SNAPSHOT_KEY = "arucon.widget.snapshot.v1"
private const val PROVIDER_CLASS = "com.arucon.widget.AruconWidgetProvider"
private const val MAXIMUM_SAFE_INTEGER = 9_007_199_254_740_991L
private val PROJECTION_KEYS = setOf(
  "petId", "stateRevision", "updatedAtMs", "formId", "personalityProfileId", "displayState"
)
private val DISPLAY_STATES = setOf("awake", "sleeping", "hibernating", "needs_care")

private fun validSnapshot(raw: String): Boolean = runCatching {
  val json = JSONObject(raw)
  val keys = mutableSetOf<String>()
  val iterator = json.keys()
  while (iterator.hasNext()) keys.add(iterator.next())
  require(keys == PROJECTION_KEYS)
  listOf("petId", "formId", "personalityProfileId").forEach { key ->
    require(json.get(key) is String && json.getString(key).isNotEmpty())
  }
  val displayState = json.get("displayState")
  require(displayState is String && displayState in DISPLAY_STATES)
  listOf("stateRevision", "updatedAtMs").forEach { key ->
    val value = json.get(key)
    require(value is Byte || value is Short || value is Int || value is Long)
    require((value as Number).toLong() in 0..MAXIMUM_SAFE_INTEGER)
  }
}.isSuccess

class AruconWidgetBridgeModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("AruconWidgetBridge")

    AsyncFunction("replaceSnapshotJsonAsync") { snapshotJson: String ->
      if (!validSnapshot(snapshotJson)) throw CodedException("Widget snapshot violates the six-field contract")
      context.getSharedPreferences(SNAPSHOT_PREFERENCES, Context.MODE_PRIVATE)
        .edit()
        .putString(SNAPSHOT_KEY, snapshotJson)
        .apply()
    }

    AsyncFunction<String?>("readSnapshotJsonAsync") {
      val raw = context.getSharedPreferences(SNAPSHOT_PREFERENCES, Context.MODE_PRIVATE)
        .getString(SNAPSHOT_KEY, null)
        ?: return@AsyncFunction null
      return@AsyncFunction if (validSnapshot(raw)) raw else null
    }

    AsyncFunction<String>("requestTimelineReloadAsync") {
      val manager = AppWidgetManager.getInstance(context)
      val provider = ComponentName(context.packageName, PROVIDER_CLASS)
      val widgetIds = manager.getAppWidgetIds(provider)
      if (widgetIds.isEmpty()) return@AsyncFunction "deferred"
      val update = Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE)
        .setComponent(provider)
        .putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, widgetIds)
      context.sendBroadcast(update)
      return@AsyncFunction "requested"
    }
  }
}
